File: apps/demo-be-elixir-phoenix/test/unit/steps/security_steps.exs
defmodule DemoBeExphWeb.Unit.SecuritySteps do
  use Cabbage.Feature, async: false, file: "security/security.feature"

  use DemoBeExphWeb.ConnCase

  alias DemoBeExph.Integration.Helpers
  alias DemoBeExph.Test.InMemoryStore

  @moduletag :unit

  defp accounts, do: Application.get_env(:demo_be_exph, :accounts_module)

  defgiven ~r/^the API is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered with password "(?<password>[^"]+)"$/,
           %{username: username, password: password},
           state do
    email = "#{username}@example.com"
    user = Helpers.register_user!(username, email, password)
    {:ok, Map.merge(state, %{alice: user, alice_password: password})}
  end

  defgiven ~r/^"alice" has had the maximum number of failed login attempts$/,
           _vars,
           %{alice: user} = state do
    InMemoryStore.update_state(fn s ->
      locked_user =
        Map.merge(user, %{
          status: "LOCKED",
          failed_login_attempts: 5,
          locked_at: DateTime.utc_now() |> DateTime.truncate(:second)
        })

      Map.update!(s, :users, fn users -> Map.put(users, user.id, locked_user) end)
    end)

    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered and locked after too many failed logins$/,
           %{username: username},
           state do
    password = "Str0ng#Pass1"
    email = "#{username}@example.com"
    user = Helpers.register_user!(username, email, password)
    {:ok, _} = accounts().deactivate_user(user)

    InMemoryStore.update_state(fn s ->
      Map.update!(s, :users, fn users ->
        Map.update!(users, user.id, fn u -> Map.put(u, :status, "LOCKED") end)
      end)
    end)

    updated_user = accounts().get_user(user.id)
    {:ok, Map.merge(state, %{alice: updated_user, alice_password: password})}
  end

  defgiven ~r/^an admin user "(?<username>[^"]+)" is registered and logged in$/,
           %{username: username},
           state do
    email = "#{username}@example.com"
    password = "Str0ng#Pass1"
    user = Helpers.register_user!(username, email, password)
    admin_user = Helpers.make_admin!(user)
    {access_token, _} = Helpers.login_user!(admin_user)
    {:ok, Map.merge(state, %{admin: admin_user, admin_token: access_token})}
  end

  defgiven ~r/^an admin has unlocked alice's account$/, _vars, %{alice: alice} = state do
    accounts().unlock_user(alice)
    {:ok, state}
  end

  defwhen ~r/^the client sends POST \/api\/v1\/auth\/register with body \{ "username": "(?<username>[^"]+)", "email": "(?<email>[^"]+)", "password": "(?<password>[^"]+)" \}$/,
          %{username: username, email: email, password: password},
          state do
    body = Jason.encode!(%{username: username, email: email, password: password})

    conn =
      build_conn()
      |> put_req_header("content-type", "application/json")
      |> post("/api/v1/auth/register", body)

    {:ok, Map.put(state, :conn, conn)}
  end

  defwhen ~r/^the client sends POST \/api\/v1\/auth\/login with body \{ "username": "(?<username>[^"]+)", "password": "(?<password>[^"]+)" \}$/,
          %{username: username, password: password},
          state do
    body = Jason.encode!(%{username: username, password: password})

    conn =
      build_conn()
      |> put_req_header("content-type", "application/json")
      |> post("/api/v1/auth/login", body)

    {:ok, Map.put(state, :conn, conn)}
  end

  defwhen ~r/^the admin sends POST \/api\/v1\/admin\/users\/\{alice_id\}\/unlock$/,
          _vars,
          %{alice: alice, admin_token: admin_token} = state do
    conn =
      build_conn()
      |> put_req_header("authorization", Helpers.bearer_header(admin_token))
      |> put_req_header("content-type", "application/json")
      |> post("/api/v1/admin/users/#{alice.id}/unlock", "{}")

    {:ok, Map.put(state, :conn, conn)}
  end

  defthen ~r/^the response status code should be (?<code>\d+)$/,
          %{code: code},
          %{conn: conn} = state do
    assert conn.status == String.to_integer(code)
    {:ok, state}
  end

  defthen ~r/^the response body should contain a validation error for "(?<field>[^"]+)"$/,
          %{field: field},
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert Map.has_key?(body, "errors")
    errors = body["errors"]
    assert Map.has_key?(errors, field)
    assert errors[field] != []
    {:ok, state}
  end

  defthen ~r/^alice's account status should be "(?<status>[^"]+)"$/,
          %{status: status},
          %{alice: alice} = state do
    updated = accounts().get_user(alice.id)
    assert String.downcase(updated.status) == String.downcase(status)
    {:ok, state}
  end

  defthen ~r/^the response body should contain a non-null "(?<field>[^"]+)" field$/,
          %{field: field},
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert Map.has_key?(body, field)
    assert body[field] != nil
    {:ok, state}
  end
end

File: apps/demo-be-e2e/tests/steps/security/security.steps.ts
import { createBdd } from "playwright-bdd";
import { setAdminRole } from "../../fixtures/db-cleanup";
import { getTokenForUser, setTokenForUser, getIdForUser, setIdForUser } from "../../utils/token-store";

const { Given } = createBdd();

async function ensureSuperadmin(request: import("@playwright/test").APIRequestContext): Promise<string> {
  try {
    return getTokenForUser("superadmin");
  } catch {
    const regRes = await request.post("/api/v1/auth/register", {
      data: { username: "superadmin", email: "superadmin@example.com", password: "Str0ng#Pass1" },
      headers: { "Content-Type": "application/json" },
    });
    const regBody = (await regRes.json()) as Record<string, unknown>;
    if (regBody["id"]) {
      setIdForUser("superadmin", regBody["id"] as string);
    }
    await setAdminRole("superadmin");
    const loginRes = await request.post("/api/v1/auth/login", {
      data: { username: "superadmin", password: "Str0ng#Pass1" },
      headers: { "Content-Type": "application/json" },
    });
    const loginBody = (await loginRes.json()) as Record<string, unknown>;
    const token = loginBody["accessToken"] as string;
    setTokenForUser("superadmin", token);
    return token;
  }
}

// ---------------------------------------------------------------------------
// Security steps: account lockout scenarios
// ---------------------------------------------------------------------------

Given("{string} has had the maximum number of failed login attempts", async ({ request }, username: string) => {
  // Trigger account lockout by sending multiple bad-password requests.
  // The backend typically locks after 5 failed attempts.
  for (let i = 0; i < 6; i++) {
    await request.post("/api/v1/auth/login", {
      data: { username, password: "WrongPassword1!" },
      headers: { "Content-Type": "application/json" },
    });
  }
});

Given(
  "a user {string} is registered and locked after too many failed logins",
  async ({ request }, username: string) => {
    // Register the user first
    const regRes = await request.post("/api/v1/auth/register", {
      data: { username, email: `${username}@example.com`, password: "Str0ng#Pass1" },
      headers: { "Content-Type": "application/json" },
    });
    const regBody = (await regRes.json()) as Record<string, unknown>;
    if (regBody["id"]) {
      setIdForUser(username, regBody["id"] as string);
    }
    // Trigger lockout with repeated bad-password attempts
    for (let i = 0; i < 6; i++) {
      await request.post("/api/v1/auth/login", {
        data: { username, password: "WrongPassword1!" },
        headers: { "Content-Type": "application/json" },
      });
    }
  },
);

Given("an admin has unlocked alice's account", async ({ request }) => {
  const adminToken = await ensureSuperadmin(request);
  const aliceId = getIdForUser("alice");
  await request.post(`/api/v1/admin/users/${aliceId}/unlock`, {
    headers: { Authorization: `Bearer ${adminToken}` },
  });
});
